soc/xtensa/intel_adsp/tools/cavstool.py

```diff
@@ -659,30 +662,62 @@ async def _main(server):
         if dsp.HIPCIDA & 0x80000000:
             dsp.HIPCIDA = 1<<31 # must ACK any DONE interrupts that arrive!
 
+        if server:
+            # Check if the client connection is alive.
+            if not is_connection_alive(server):
+                lock.acquire()
+                start_output = False
+                lock.release()
+
 class adsp_request_handler(socketserver.BaseRequestHandler):
     """
     The request handler class for control the actions of server.
     """
 
-    def receive_fw(self, filename):
+    def receive_fw(self):
+        header_sz = 80
+        LARGE_BUF_SIZE = 16 * 1024 * 1024
+
+        log.info("Receiving...")
+        # The first contain the header and the front
+        # data of the firmware
+        d = self.request.recv(BUF_SIZE)
+
+        # Unpacked the header data
+        # Include size(4), filename(42), MD5(32) and pad(2)
+        header = d[:header_sz]
+        total = d[header_sz:]
+        s = struct.Struct(PACKET_HEADER_FORMAT_FW)
+        fsize, fname, md5_tx_b, pad = s.unpack(header)
+        log.info(f'size:{fsize}, filename:{fname}, MD5:{md5_tx_b}, pad:{pad})')
+
+        # Receive the remain part of the firmware. We only
+        # receive the specified amount of bytes.
+        while len(total) < fsize:
+            data = self.request.recv(LARGE_BUF_SIZE)
```

Comment:
```suggestion
            data = self.request.recv(min(LARGE_BUF_SIZE, fsize - len(total))
```
otherwise we're back to the bug https://github.com/zephyrproject-rtos/zephyr/issues/46864 that this is supposed to fix.

From https://docs.python.org/3/howto/sockets.html
> One complication to be aware of: if your conversational protocol allows multiple messages to be sent back to back (without some kind of reply), and you pass recv an arbitrary chunk size, you may end up reading the start of a following message

Reply:
Good suggestion, applied it.